### ISSUE TITLE
Metrics calculation for ASSET TEST Gold Reference

### ISSUE THREAD

**Issue by pelican9** [facebookresearch/muss#23]
Hi,
May I know how the metrics for ASSET TEST gold reference is calculated? I tried to do it in a leave-one- out scenario where each reference is evaluated against all others but the SARI is a little bit different to the result listed in the paper but FKGL is the same. It would be great if I can know what is wrong with my understanding. Following is my code
```
complete_ref_path = ''
for i in range(10):
  complete_ref_path += '/content/drive/MyDrive/muss/resources/datasets/asset/test.simple.'+str(i)
  if i != 9:
    complete_ref_path += ','
complete_ref_path_list = complete_ref_path.split(',')
complete_ref_path_list
```
which gives
```
['/content/drive/MyDrive/muss/resources/datasets/asset/test.simple.0',
 '/content/drive/MyDrive/muss/resources/datasets/asset/test.simple.1',
 '/content/drive/MyDrive/muss/resources/datasets/asset/test.simple.2',
 '/content/drive/MyDrive/muss/resources/datasets/asset/test.simple.3',
 '/content/drive/MyDrive/muss/resources/datasets/asset/test.simple.4',
 '/content/drive/MyDrive/muss/resources/datasets/asset/test.simple.5',
 '/content/drive/MyDrive/muss/resources/datasets/asset/test.simple.6',
 '/content/drive/MyDrive/muss/resources/datasets/asset/test.simple.7',
 '/content/drive/MyDrive/muss/resources/datasets/asset/test.simple.8',
 '/content/drive/MyDrive/muss/resources/datasets/asset/test.simple.9']
```

```
from easse.cli import evaluate_system_output

test_set = 'custom'

bleu_list,fkgl_list,sari_list, sari_add_list,sari_del_list,sari_keep_list= [],[],[],[],[],[]

for i in range(10):

  sys_sents_path = complete_ref_path_list[i]
  orig_sents_path = '/content/drive/MyDrive/muss/resources/datasets/asset/test.complex'
  refs_sents_paths = ','.join(complete_ref_path_list[:i]+complete_ref_path_list[i+1:])

  result = evaluate_system_output(
      test_set,
      sys_sents_path=sys_sents_path,
      orig_sents_path=orig_sents_path,
      refs_sents_paths=refs_sents_paths,
      metrics=['bleu','fkgl','sari', 'sari_by_operation'],
      quality_estimation=False,
  )

  bleu_list.append(result['bleu'])
  fkgl_list.append(result['fkgl'])
  sari_list.append(result['sari'])
  sari_add_list.append(result['sari_add'])
  sari_del_list.append(result['sari_del'])
  sari_keep_list.append(result['sari_keep'])

```
And the result is 
```
np.mean(sari_list),get_mean_confidence_interval(sari_list)
(44.88725475437498, 0.32906356454497476)

np.mean(fkgl_list),get_mean_confidence_interval(fkgl_list)
(6.487483272312891, 0.15439073457017022)
```

Thank you in advance!

**Comment by louismartin** [comment]
Hi @pelican9 that's  a good question!
The code that I used to do this computation is here: https://github.com/facebookresearch/muss/blob/8b016942d9d0984866f6832594c049a189c1b46e/muss/mining/training.py#L353

I think the difference is that I might have duplicated one of the references so that the method does not change the total number of references used.
Does that make sense?

**Comment by pelican9** [comment]
Hi, @louismartin Yes, it makes sense to me. Thank you for providing the location of the source code. It's exactly what I wanted!

If you don't mind, may I ask another question about ACCESS? In the paper _Controllable sentence simplification_, it is mentioned that "Special tokens are then included in the vocabulary (40 unique values per parameter token)." Would you mind advising me where I can find these tokens as vocabulary? I looked at https://github.com/facebookresearch/access/blob/7b61fbf0bad665798d662e0a90d2a0e451367df6/resources/various/sentencepiece_model/sentencepiece_model_10000.vocab but it seems they are not there. And when I do the following, it seems each control token is broke into small chunks instead of a whole token. I suppose for each token with a specified value, it should be encoded into only 1 token instead of a few?

```
from muss.preprocessors import GPT2BPEPreprocessor

text = '<DEPENDENCYTREEDEPTHRATIO_0.9> <WORDRANKRATIO_1.05> <REPLACEONLYLEVENSHTEIN_0.35> <LENGTHRATIO_1.05> Upper Sorbian is a minority language spoken by Sorbs in Germany in the historical province of Upper Lusatia (Hornja Å uÅ 3\/4 ica in Sorbian), which is today part of Saxony.'

gpt2 = GPT2BPEPreprocessor()
encode = gpt2.encode_sentence(text)
encode
# '27 46162 10619 45155 51 2200 1961 8905 4221 49 1404 9399 62 15 13 24 29 1279 45359 7707 15154 49 1404 9399 62 16 13 2713 29 1279 2200 6489 11598 1340 11319 2538 53 16938 6535 36 1268 62 15 13 2327 29 1279 43 49494 49 1404 9399 62 16 13 2713 29 20390 15423 12210 318 257 9137 3303 9635 416 15423 1443 287 4486 287 262 6754 8473 286 20390 406 37937 544 357 39 1211 6592 6184 227 334 127 227 513 11139 19 220 3970 287 15423 12210 828 543 318 1909 636 286 29242 1647 13'

gpt2.decode_sentence('27 46162 10619')
# '<DEPEND'

##############
# What I thought is something like
gpt2.encode_sentence('<DEPENDENCYTREEDEPTHRATIO_0.9>')
# 99999

gpt2.decode_sentence('99999')
# '<DEPENDENCYTREEDEPTHRATIO_0.9>'
```

Thank you in advance!

**Comment by louismartin** [comment]
Hum that's a good point, for the original ACCESS, the special tokens were added in the fairseq vocabulary only ([see here](http://dl.fbaipublicfiles.com/access/best_model.tar.gz)).
Sentence piece is only used to tokenize the raw text and it isn't applied on the special tokens.


Related to MUSS, as we used the pretrained model BART, we couldn't modify the vocabulary, so indeed the special tokens are split into multiple subtokens: e.g. `<WORDRANKRATIO_1.05>` will be split in something like ` ['<W', 'OR', 'DRA', ...]`.
I know it's not super clean but in my experiments it didn't make a difference in performance so I stayed with this simpler method.
